File: static/resume-builder.js
'use strict';
const form = document.querySelector('#resume-form');
const profileSelect = document.querySelector('#resume-profile');
const statusMessage = document.querySelector('#builder-status');
const fields = ['name', 'email', 'phone', 'location', 'headline', 'summary', 'skills', 'experience', 'education', 'certifications'];
let workspace;
let source;
let draftKey;
let finalResult = null;
let pendingTask = null;
let tailoringBusy = false;
let aiAvailable = null;
const tailorForm = document.querySelector('#tailor-form');
const tailorStatus = document.querySelector('#tailor-status');
const downloadButtons = document.querySelectorAll('[data-download]');
const downloadStatus = document.querySelector('#download-status');
function setDownloadsDisabled(disabled) {
    downloadButtons.forEach(button => { button.disabled = disabled || tailoringBusy; });
}

function values() {
    return Object.fromEntries(fields.map(key => [key, form.elements[key].value]));
}
function resumeText() {
    if (finalResult) return finalResult.text;
    const data = values();
    const lines = [data.name.trim(), data.headline.trim(), [data.email, data.phone, data.location].map(s => s.trim()).filter(Boolean).join(' | ')].filter(Boolean);
    for (const [key, title] of [['summary', 'PROFESSIONAL SUMMARY'], ['skills', 'SKILLS'], ['experience', 'WORK EXPERIENCE'], ['education', 'EDUCATION'], ['certifications', 'CERTIFICATIONS']]) {
        if (data[key].trim()) lines.push('', title, data[key].trim());
    }
    return lines.join('\n');
}
function preview() {
    document.querySelector('#resume-preview').textContent = resumeText() || 'Add your details to start building your resume.';
    document.querySelector('#preview-heading').textContent = finalResult ? 'Final tailored resume' : 'Live preview';
    document.querySelector('#download-description').textContent = finalResult ? 'Download the final tailored resume shown below as PDF, Word or text.' : 'Download your current edits as a PDF, Word document or plain text.';
}
function saveDraft() {
    preview();
    try {
        localStorage.setItem(draftKey, JSON.stringify(values()));
        statusMessage.textContent = `Draft saved for ${workspace.profile.title}.`;
    } catch {
        statusMessage.textContent = 'Browser storage is unavailable. Download your resume to keep your edits.';
    }
}
function populate(data) {
    fields.forEach(key => { form.elements[key].value = typeof data[key] === 'string' ? data[key] : ''; });
    preview();
}
async function loadProfile() {
    const response = await fetch('/api/state');
    if (!response.ok) throw Error('Could not load your profile. Reload this page to try again.');
    workspace = await response.json();
    source = {...workspace.profile, certifications: workspace.combined_certifications ?? workspace.profile.certifications};
    renderApplicationContext();
    draftKey = `jobflow.resume.${source.id}`;
    profileSelect.replaceChildren(...workspace.profiles.map(p => new Option(p.title, p.id)));
    profileSelect.value = source.id;
    let draft;
    let storageUnavailable = false;
    try { draft = JSON.parse(localStorage.getItem(draftKey)); } catch { storageUnavailable = true; }
    const restored = draft && typeof draft === 'object' && !Array.isArray(draft);
    populate(restored ? draft : source);
    restoreTailoring();
    statusMessage.textContent = storageUnavailable ? 'Browser storage is unavailable. Download your resume to keep your edits.' : restored ? `Saved draft restored for ${source.title}. Use “Refill from saved profile” to load updated profile information.` : `Prefilled from ${source.title}. Add or edit any missing details below.`;
    document.querySelector('#resume-fields').disabled = false;
    setDownloadsDisabled(false);
    document.querySelector('#tailor-fields').disabled = false;
    if (pendingTask) pollTask(pendingTask);
}
form.addEventListener('input', event => {
    if (event.target !== profileSelect) { invalidateTailoring(); saveDraft(); }
});
document.querySelector('#resume-reset').onclick = async () => {
    if (!confirm('Replace this resume draft with the latest saved profile information?')) return;
    try {
        const response = await fetch('/api/state');
        if (!response.ok) throw Error('Could not load the latest profile. Try again.');
        const latest = await response.json();
        if (latest.profile.id !== source.id) throw Error('The active profile changed in another page. Reload this page first.');
        source = {...latest.profile, certifications: latest.combined_certifications ?? latest.profile.certifications};
        workspace.profile = latest.profile;
        renderApplicationContext();
        invalidateTailoring();
        populate(source);
        saveDraft();
        saveTailoring();
    } catch (error) { statusMessage.textContent = error.message; }
};
profileSelect.onchange = async () => {
    document.querySelector('#resume-fields').disabled = true;
    document.querySelector('#tailor-fields').disabled = true;
    setDownloadsDisabled(true);
    try {
        const response = await fetch('/api/profiles/select', {method: 'POST', headers: {'Content-Type': 'application/json', 'X-Session-Token': workspace.token}, body: JSON.stringify({id: profileSelect.value})});
        if (!response.ok) throw Error((await response.json()).error);
        await loadProfile();
    } catch (error) {
        profileSelect.value = source.id;
        statusMessage.textContent = error.message;
    } finally {
        document.querySelector('#resume-fields').disabled = tailoringBusy;
        document.querySelector('#tailor-fields').disabled = tailoringBusy;
        setDownloadsDisabled(false);
    }
};
form.onsubmit = async event => {
    event.preventDefault();
    const format = event.submitter?.dataset.download || 'docx';
    setDownloadsDisabled(true);
    downloadStatus.textContent = 'Preparing your download…';
    try {
        const response = await fetch('/api/resume-builder/download', {method: 'POST', headers: {'Content-Type': 'application/json', 'X-Session-Token': workspace.token}, body: JSON.stringify({text: resumeText(), format})});
        if (!response.ok) throw Error((await response.json()).error);
        const url = URL.createObjectURL(await response.blob());
        const link = document.createElement('a');
        link.href = url;
        link.download = `resume.${format}`;
        document.body.append(link);
        link.click();
        link.remove();
        setTimeout(() => URL.revokeObjectURL(url), 10000);
        downloadStatus.textContent = `Your ${{docx: 'Word', pdf: 'PDF', txt: 'text'}[format]} resume download has started.`;
    } catch (error) { downloadStatus.textContent = error.message; }
    finally { setDownloadsDisabled(false); }
};
function jobValues() {
    return {title: document.querySelector('#target-title').value,
        description: document.querySelector('#target-description').value,
        engine: document.querySelector('#tailor-engine').value,
        preferences: {length: document.querySelector('#resume-length').value,
            tone: document.querySelector('#resume-tone').value,
            emphasis: document.querySelector('#resume-emphasis').value,
            max_skills: Number(document.querySelector('#resume-max-skills').value)}};
}
function savedContext() {
    const p = workspace?.profile || {};
    return {roles: p.roles || '', search_location: p.search_location || '', work_rights: p.work_rights || '', constraints: p.constraints || '', answers: p.answers || {}};
}
function renderApplicationContext() {
    const p = savedContext();
    document.querySelector('#analysis-context').textContent = [
        `Target roles: ${p.roles || 'Not provided'}`, `Search area: ${p.search_location || 'Not provided'}`,
        `Work rights: ${p.work_rights || 'Not provided'}`, `Availability and other constraints: ${p.constraints || 'Not provided'}`,
        `Saved application answers: ${JSON.stringify(p.answers, null, 2)}`
    ].join('\n\n');
}
function fingerprint() { return JSON.stringify({candidate: values(), context: savedContext(), ...jobValues()}); }
function saveTailoring() {
    try {
        localStorage.setItem(`${draftKey}.tailoring`, JSON.stringify({...jobValues(), result: finalResult, task: pendingTask, fingerprint: fingerprint()}));
    } catch {
        tailorStatus.textContent = 'Browser storage is unavailable. Keep this page open and download your final resume.';
    }
}
function renderAnalysis() {
    document.querySelector('#tailor-results').hidden = !finalResult;
    document.querySelector('#final-edit').hidden = !finalResult;
    document.querySelector('#final-resume-text').value = finalResult?.text || '';
    if (!finalResult) { preview(); return; }
    const labels = {evidenced: 'Evidence found — verify', partial: 'Partial evidence', not_found: 'Not evidenced in profile', review: 'Check wording overlap'};
    document.querySelector('#analysis-note').textContent = `${finalResult.engine === 'ollama' ? 'Local AI analysis.' : 'Basic wording analysis; it cannot verify eligibility.'} ${finalResult.note}`;
    const preferences = finalResult.preferences;
    document.querySelector('#analysis-settings').textContent = preferences
        ? `${finalResult.model || 'Basic matching'} · ${preferences.length} detail · ${preferences.tone} style · ${preferences.emphasis.replaceAll('_', ' ')} · Up to ${preferences.max_skills} skills`
        : '';
    document.querySelector('#analysis-skills').textContent = finalResult.skills.join(', ') || 'No relevant saved skills identified.';
    const tbody = document.querySelector('#analysis-requirements');
    tbody.replaceChildren();
    for (const requirement of finalResult.requirements) {
        const row = document.createElement('tr');
        const target = document.createElement('td');
        target.textContent = requirement.requirement;
        const evidence = document.createElement('td');
        evidence.textContent = requirement.evidence.length ? requirement.evidence.join('\n\n') : 'No supporting evidence identified. Add details only if true.';
        const review = document.createElement('td');
        const label = document.createElement('strong');
        label.textContent = labels[requirement.status] || 'Needs review';
        const note = document.createElement('p');
        note.textContent = requirement.note;
        review.append(label, note);
        row.append(target, evidence, review);
        tbody.append(row);
    }
    for (const [selector, items] of [['#analysis-warnings', [...finalResult.warnings, 'Verify the generated summary, selected experience, credential validity and all job requirements before applying.']], ['#analysis-ats', finalResult.ats]]) {
        document.querySelector(selector).replaceChildren(...items.map(item => {
            const li = document.createElement('li'); li.textContent = item; return li;
        }));
    }
    preview();
}
function invalidateTailoring() {
    if (finalResult) tailorStatus.textContent = 'Your details changed. Run analysis again to create an updated tailored resume.';
    finalResult = null;
    pendingTask = null;
    downloadStatus.textContent = '';
    updateEngineNote();
    renderAnalysis();
    saveTailoring();
}
function restoreTailoring() {
    finalResult = null;
    pendingTask = null;
    tailorStatus.textContent = '';
    let saved;
    try { saved = JSON.parse(localStorage.getItem(`${draftKey}.tailoring`)); } catch { /* Keep source form available. */ }
    document.querySelector('#target-title').value = typeof saved?.title === 'string' ? saved.title : '';
    document.querySelector('#target-description').value = typeof saved?.description === 'string' ? saved.description : '';
    document.querySelector('#tailor-engine').value = saved?.engine === 'basic' ? 'basic' : 'ollama';
    for (const [id, key, fallback] of [['resume-length', 'length', 'balanced'], ['resume-tone', 'tone', 'direct'], ['resume-emphasis', 'emphasis', 'role_fit'], ['resume-max-skills', 'max_skills', '12']]) {
        const select = document.getElementById(id);
        const value = String(saved?.preferences?.[key] ?? fallback);
        select.value = [...select.options].some(option => option.value === value) ? value : fallback;
    }
    updateEngineNote();
    if (saved?.fingerprint === fingerprint()) {
        if (saved.result && typeof saved.result.text === 'string' && ['skills', 'requirements', 'warnings', 'ats'].every(key => Array.isArray(saved.result[key]))) finalResult = saved.result;
        if (typeof saved.task === 'string') pendingTask = saved.task;
    }
    renderAnalysis();
    if (finalResult) tailorStatus.textContent = 'Your saved tailored resume is ready to review and download.';
}
function setTailoringBusy(busy) {
    tailoringBusy = busy;
    document.querySelector('#resume-fields').disabled = busy;
    document.querySelector('#tailor-fields').disabled = busy;
    document.querySelector('#resume-use-source').disabled = busy;
    document.querySelector('#final-resume-text').disabled = busy;
    document.querySelector('#tailor-submit').textContent = busy ? 'Analyzing and tailoring…' : 'Analyze job & tailor resume';
    setDownloadsDisabled(busy);
}
async function pollTask(id) {
    setTailoringBusy(true);
    try {
        while (true) {
            const response = await fetch(`/api/resume-builder/tasks/${encodeURIComponent(id)}`);
            const task = await response.json();
            if (!response.ok) throw Error(task.error);
            tailorStatus.textContent = task.message;
            if (task.status === 'error') throw Error(task.message);
            if (task.status === 'complete') {
                finalResult = task.result;
                pendingTask = null;
                saveTailoring();
                renderAnalysis();
                return;
            }
            await new Promise(resolve => setTimeout(resolve, 1500));
        }
    } catch (error) {
        pendingTask = null;
        saveTailoring();
        tailorStatus.textContent = `${error.message} Your source information and any previous final resume are preserved.`;
    } finally { setTailoringBusy(false); }
}
tailorForm.addEventListener('input', invalidateTailoring);
tailorForm.onsubmit = async event => {
    event.preventDefault();
    if (!form.reportValidity()) return;
    setTailoringBusy(true);
    tailorStatus.textContent = 'Starting analysis. Local AI can take a few minutes; your progress is saved if you reload.';
    try {
        const response = await fetch('/api/resume-builder/tailor', {method: 'POST', headers: {'Content-Type': 'application/json', 'X-Session-Token': workspace.token},
            body: JSON.stringify({profile_id: source.id, candidate: values(), ...jobValues()})});
        const task = await response.json();
        if (!response.ok) throw Error(task.error);
        pendingTask = task.id;
        saveTailoring();
        await pollTask(task.id);
    } catch (error) {
        tailorStatus.textContent = error.message;
    } finally { setTailoringBusy(false); }
};
document.querySelector('#resume-use-source').onclick = () => {
    invalidateTailoring();
    tailorStatus.textContent = 'Using your untailored source resume. Your job description is saved for the next analysis.';
};
document.querySelector('#final-resume-text').oninput = event => {
    if (!finalResult) return;
    finalResult.text = event.target.value;
    preview();
    saveTailoring();
    tailorStatus.textContent = 'Final resume edits saved. Review your changes before downloading.';
};
loadProfile().catch(error => { statusMessage.textContent = error.message; });
function updateEngineNote() {
    const basic = document.querySelector('#tailor-engine').value === 'basic';
    for (const id of ['resume-length', 'resume-tone', 'resume-emphasis']) document.getElementById(id).disabled = basic;
    document.querySelector('#tailor-engine-note').textContent = document.querySelector('#tailor-engine').value === 'basic'
        ? 'Basic matching selects source facts using wording overlap. It runs without AI and cannot assess eligibility.'
        : aiAvailable === true ? 'Local Qwen AI is ready. Analysis and rewriting stay on this computer.'
        : aiAvailable === false ? 'Local Qwen AI is unavailable. Start Ollama with qwen3:8b, or select Basic matching to continue without AI.'
        : 'Local AI uses Qwen on this computer. Checking availability…';
}
async function checkLocalModel() {
    const button = document.querySelector('#refresh-model');
    const status = document.querySelector('#local-model-status');
    button.disabled = true;
    status.textContent = 'Checking local model...';
    try {
        const response = await fetch('/api/ai');
        if (!response.ok) throw Error('Could not check the local model. Try again.');
        const ai = await response.json();
        aiAvailable = Boolean(ai.available);
        status.textContent = aiAvailable ? `${ai.model} is ready through Ollama. Your information stays on this computer.`
            : `${ai.model || 'qwen3:8b'} is not ready. Start Ollama and run: ollama pull qwen3:8b. Then check the connection again.`;
    } catch (error) { aiAvailable = false; status.textContent = error.message; }
    finally { button.disabled = false; updateEngineNote(); }
}
document.querySelector('#refresh-model').onclick = checkLocalModel;
checkLocalModel();

File: resume_builder.py
"""Background, profile-scoped resume analysis using the existing local tailor."""
import copy
import json
import re
import threading
import time
import uuid

from tailoring import confirmed_fact, focused_profile, relevant_fact, requirements_text, sections

FIELDS = ('name', 'email', 'phone', 'location', 'headline', 'summary', 'skills', 'experience', 'education', 'certifications')
TASKS = {}
LOCK = threading.Lock()


def normalized(text):
    return ' '.join(text.casefold().split())


def application_context(profile, candidate):
    context = {key: str(profile.get(key, '') or '') for key in ('roles', 'search_location', 'work_rights', 'constraints')}
    context['location'] = candidate.get('location', '')
    context['saved_application_answers'] = json.dumps(profile.get('answers', {}), ensure_ascii=False)
    return context


def analysis(profile, job, engine, preferences=None):
    """Every displayed requirement and piece of evidence must be traceable."""
    description = requirements_text(job['description'])
    facts = {key: profile[key] for key in ('headline', 'summary', 'skills', 'experience', 'education', 'certifications')}
    facts.update(profile.get('_application_context', {}))
    if engine == 'ollama':
        from local_ai import structured
        # Fit analysis needs the full advertisement, including salary/roster
        # details that may appear alongside benefits or company information.
        description = job['description']
        item = {'type': 'object', 'properties': {
            'requirement': {'type': 'string'},
            'evidence': {'type': 'array', 'items': {'type': 'string'}, 'maxItems': 3},
            'status': {'type': 'string', 'enum': ['evidenced', 'partial', 'not_found']},
            'note': {'type': 'string'},
        }, 'required': ['requirement', 'evidence', 'status', 'note'], 'additionalProperties': False}
        result = structured(
            'Read the complete advertisement. Distinguish actual applicant requirements and working conditions '
            'from company marketing or benefits; do not treat company capabilities as applicant skills. '
            'Compare the main responsibilities, essential and preferred criteria, tools, qualifications, '
            'experience, and working conditions in the job against the candidate facts. Return up to 20 '
            'distinct requirements, retaining mandatory/preferred conditions. Each requirement must be '
            'an EXACT quotation from the description. Evidence must be short EXACT quotations from '
            'candidate fields, never from the job. Use evidenced only for explicit supporting facts; '
            'partial for transferable or incomplete evidence; not_found for absent or uncertain evidence. '
            'An expired credential does not satisfy a current requirement. Do not infer qualifications '
            'from generic duties or count negated/conditional facts as evidence. Briefly explain gaps '
            'and what factual information the applicant should verify. Check work rights, location, availability, '
            'salary or roster constraints and saved screening answers when the job specifies them. A target role '
            'is a preference, not evidence of experience; a home address does not prove willingness to relocate. '
            'Respect limitations in application answers. No scores or hiring predictions.',
            {'candidate': facts, 'job': {**job, 'description': description}, 'writing_preferences': preferences or {}},
            {'type': 'object', 'properties': {'requirements': {'type': 'array', 'items': item, 'minItems': 1, 'maxItems': 20}},
             'required': ['requirements'], 'additionalProperties': False}, max_tokens=3200)
        rows = result.get('requirements')
        if not isinstance(rows, list) or not 1 <= len(rows) <= 20:
            raise ValueError('The requirement analysis was incomplete. Try again or choose Basic matching.')
        verified = []
        for row in rows:
            if not isinstance(row, dict):
                raise ValueError('Invalid requirement analysis. Try again.')
            requirement, evidence = row.get('requirement'), row.get('evidence')
            status, note = row.get('status'), row.get('note')
            if (not isinstance(requirement, str) or not 8 <= len(requirement) <= 1200
                    or normalized(requirement) not in normalized(description)
                    or not isinstance(evidence, list) or len(evidence) > 3
                    or status not in ('evidenced', 'partial', 'not_found')
                    or not isinstance(note, str) or len(note) > 1500):
                raise ValueError('The analysis could not be traced to the job description. Try again.')
            # Keep the surrounding source line so a short quote cannot hide
            # "not yet obtained", expiry dates or other credential conditions.
            supported = []
            for quote in evidence:
                if not isinstance(quote, str) or not quote.strip():
                    continue
                for value in facts.values():
                    context = next((line.strip() for line in value.splitlines()
                                    if normalized(quote) in normalized(line) and confirmed_fact(line)), None)
                    if context:
                        supported.append(context)
                        break
            if len(supported) != len(evidence) or (status != 'not_found' and not supported):
                status, supported = 'not_found', []
                note = 'The proposed evidence could not be verified. Check this requirement against your real experience.'
            verified.append({'requirement': requirement, 'evidence': supported, 'status': status, 'note': note})
        return verified

    # Basic mode reports wording overlap, never claims that a qualification is met.
    lines = [line.strip(' \t-*•') for line in re.split(r'\n+|(?<=[.!?])\s+', description)]
    requirements = list(dict.fromkeys(line for line in lines if len(line) >= 12))[:20]
    evidence_lines = [line.strip() for key, value in facts.items()
                      for line in (re.split(r'\n|,', value) if key == 'skills' else value.splitlines()) if line.strip()]
    return [{'requirement': line,
             'evidence': [fact for fact in evidence_lines if confirmed_fact(fact) and relevant_fact(fact, {'description': line})][:3],
             'status': 'review',
             'note': 'Wording overlap only. Verify the full requirement, dates and conditions.'}
            for line in requirements]


def basic_draft(profile, job):
    """Extract relevant source facts; leave prose and achievements uninvented."""
    from local_ai import factual_summary
    focused_job = {**job, 'description': requirements_text(job['description'])}
    selection = {key: [i for i, line in enumerate(lines) if confirmed_fact(line) and relevant_fact(line, focused_job)]
                 for key, lines in sections(profile).items()}
    # Keep issuer/date/condition lines attached to selected credentials.
    for key in ('education', 'certifications'):
        for i, line in enumerate(sections(profile)[key]):
            if i - 1 in selection[key] and re.match(r'^\s*(issued|issuer|expiry|expires|awarded|valid|condition|\d{4})\b', line, re.I):
                selection[key].append(i)
    selection['headline'] = bool(profile['headline'] and relevant_fact(profile['headline'], focused_job))
    skills = list(dict.fromkeys(skill.strip() for skill in re.split(r'[,\n]', profile['skills'])
                               if skill.strip() and confirmed_fact(skill) and relevant_fact(skill, focused_job)))
    draft = {'summary': '', 'skills': skills, 'selection': selection, 'target_job': job, 'job_priorities': []}
    draft['summary'] = factual_summary(focused_profile(profile, draft))
    return draft


def generate(profile, job, engine, progress=lambda message: None, preferences=None):
    import app
    from local_ai import rewrite, resume_preferences, MODEL
    settings = resume_preferences(preferences)
    progress('Analyzing job requirements and checking your evidence…')
    requirements = analysis(profile, job, engine, settings)
    progress('Selecting relevant facts and tailoring your summary…')
    draft = rewrite(profile, job, preferences=settings) if engine == 'ollama' else basic_draft(profile, job)
    draft['skills'] = draft['skills'][:settings['max_skills']]
    focused = focused_profile(profile, draft)
    resume = app.tailor(profile, job, draft)
    warnings = []
    if engine == 'basic':
        warnings.append('Basic matching applies the skills limit. Writing style, length and emphasis require Local AI.')
    if not focused['experience'].strip():
        warnings.append('No relevant work history was selected. Add genuine transferable duties to your source information before using this resume.')
    if not draft['skills']:
        warnings.append('No relevant saved skills were found. Add relevant skills only if you actually have them.')
    if not profile['email'].strip() and not profile['phone'].strip():
        warnings.append('Add an email address or phone number so employers can contact you.')
    if not re.search(r'\d', focused['experience']):
        warnings.append('Check that each role includes its employer, title and dates. Add verified achievements where available.')
    return {'text': resume['text'], 'requirements': requirements, 'skills': draft['skills'],
            'warnings': warnings, 'engine': engine, 'model': MODEL if engine == 'ollama' else None,
            'preferences': settings,
            'reviewed_inputs': ['Job duties and selection criteria', 'Profile summary, skills and work history',
                                'Education and certifications', 'Location, target roles and search area',
                                'Saved work rights, availability constraints and application answers'],
            'note': 'Review the summary and requirement analysis for accuracy. Missing requirements are not added to your resume.',
            'ats': ['Single column with standard section headings', 'Selectable text in PDF and Word',
                    'Original employers, role titles and dates retained for selected experience',
                    'No tables, images, text boxes or keyword stuffing']}


def start(body, active_profile):
    import app
    from local_ai import resume_preferences
    settings = resume_preferences(body.get('preferences'))
    if body.get('profile_id') != active_profile['id']:
        raise ValueError('The active profile changed. Reload the builder before tailoring.')
    engine = body.get('engine', 'ollama')
    if engine not in ('ollama', 'basic'):
        raise ValueError('Choose Local AI or Basic matching.')
    candidate = body.get('candidate')
    if not isinstance(candidate, dict) or any(not isinstance(candidate.get(key, ''), str) for key in FIELDS):
        raise ValueError('Resume fields must contain text.')
    candidate = {key: candidate.get(key, '').strip() for key in FIELDS}
    context = application_context(active_profile, candidate)
    title, description = body.get('title', ''), body.get('description', '')
    if not isinstance(title, str) or not isinstance(description, str) or not 40 <= len(description.strip()) <= 16000 or len(title) > 200:
        raise ValueError('Paste a job description between 40 and 16,000 characters and a title under 200 characters.')
    if sum(len(value) for value in candidate.values()) + len(description) + len(json.dumps(context)) > 22000:
        raise ValueError('The resume, saved application context and job description are too long. Keep their combined length under 22,000 characters.')
    if not candidate['name'] or not any(candidate[key] for key in ('experience', 'education', 'skills')):
        raise ValueError('Add your name and real experience, education or skills before tailoring.')
    if not requirements_text(description):
        raise ValueError('Add job duties and requirements, not only company information or benefits.')
    candidate['_application_context'] = context
    if app.RUN_LOCK.locked():
        raise ValueError('Wait for the current browser run to finish before tailoring.')
    if not app.AI_LOCK.acquire(blocking=False):
        raise ValueError('Another analysis is running. Wait for it to finish and try again.')
    task_id = uuid.uuid4().hex
    with LOCK:
        for old_id, task in list(TASKS.items()):
            if time.monotonic() - task['created'] > 3600 or (len(TASKS) >= 20 and task['status'] != 'running'):
                del TASKS[old_id]
        TASKS[task_id] = {'id': task_id, 'profile_id': active_profile['id'], 'created': time.monotonic(),
                          'status': 'running', 'message': 'Starting resume analysis…'}

    def update(**changes):
        with LOCK:
            TASKS[task_id].update(changes)

    def work():
        try:
            result = generate(candidate, {'title': title.strip(), 'description': description.strip()}, engine,
                              lambda message: update(message=message), preferences=settings)
            update(status='complete', result=result, message='Your tailored resume is ready to review.')
        except Exception as exc:
            update(status='error', message=str(exc) if isinstance(exc, ValueError) else 'Tailoring failed. Try again or choose Basic matching.')
        finally:
            app.AI_LOCK.release()

    try:
        threading.Thread(target=work, daemon=True).start()
    except Exception:
        app.AI_LOCK.release()
        with LOCK:
            del TASKS[task_id]
        raise
    return {'id': task_id, 'status': 'running'}


def get(task_id, profile_id):
    with LOCK:
        task = TASKS.get(task_id)
        if not task or task['profile_id'] != profile_id:
            raise ValueError('Analysis not found for this profile. Run the analysis again.')
        return copy.deepcopy({key: value for key, value in task.items() if key != 'created'})
